### PR TITLE
Error messaging for unsupported ciphers

### PR DESCRIPTION
**Pull request by mpmcroy**
Adds checks in two places to prevent adding/connecting to servers with unsupported ciphers.

1) Prevent users from adding bad servers to their server list. Adds a check when users tries to add an access key and shows an error message if it has an unsupported cipher.

Screen Shot 2020-05-01 at 3 09 23 PM

2) Check servers already in server list for any that use unsupported ciphers. Added a warning message and disabled the "connect" button. This is a corner case that will decrease in numbers as time goes on since we'll be preventing invalid servers from entering the list.

Screen Shot 2020-05-01 at 11 10 01 AM

### CODE REVIEW

File: cordova-plugin-outline/outlinePlugin.js [hunk 1]
```diff
@@ -56,7 +56,8 @@ const ERROR_CODE = {
   CONFIGURE_SYSTEM_PROXY_FAILURE: 9,
   NO_ADMIN_PERMISSIONS: 10,
   UNSUPPORTED_ROUTING_TABLE: 11,
-  SYSTEM_MISCONFIGURED: 12
+  SYSTEM_MISCONFIGURED: 12,
+  UNSUPPORTED_CIPHER: 13,
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply cordova-plugin-outline/outlinePlugin.js:60]: Done

File: src/www/ui_components/server-card.html [hunk 2]
```diff
@@ -185,13 +198,17 @@
           type: String,
           computed: "_computeStatusMessage(state, localize)",
         },
+        cipherSupported: {
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: 👍

File: src/www/ui_components/server-card.html [hunk 3]
```diff
@@ -225,8 +242,11 @@
           }[state] || this.localize("connect-button-label")
         );
       },
-      _computeConnectButtonDisabled: function(state) {
-        return state === "CONNECTING" || state === "DISCONNECTING";
+      _computeConnectButtonDisabled: function(state, cipherSupported) {
+        return !cipherSupported || state === "CONNECTING" || state === "DISCONNECTING";
+      },
+      _isCipherSupported: function(cipher) {
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: 👍

File: src/www/ui_components/server-card.html [hunk 4]
```diff
@@ -169,6 +177,10 @@
     Polymer({
       is: "server-card",
       properties: {
+        SUPPORTED_CIPHERS: {
```
Comment (fortuna) [inline src/www/ui_components/server-card.html:180]: The UI element shouldn't know about what ciphers are supported.
Instead, expose only isCipherSupported, and let the caller set that.

Provide a method isCipherSupported(cipher), rather than a list. It's more flexible and extensible. The caller can use that method to set the field here.

That will also remove the duplicate list.

File: src/www/messages/en.json [hunk 5]
```diff
@@ -19,6 +19,7 @@
   "error-server-incompatible": "Sorry, this access key is not compatible with this version of Outline.",
   "error-timeout": "Something seems to be taking longer than expected. Quitting and restarting may help. If this happens again, please submit feedback.",
   "error-unexpected": "Sorry, an unexpected error occurred. Quitting and restarting may help. If this happens again, please submit feedback.",
+  "error-unsupported-cipher": "Sorry, the server you are trying to connect to uses an unsupported cipher. Non-AEAD ciphers are not supported.",
```
Comment (fortuna) [inline src/www/messages/en.json:22]: We need a more actionable error. Maybe something like "The server you added uses cipher {cipher}, but Outline only supports AEAD ciphers. Please contact the admin of the server."

File: src/www/app/persistent_server.ts [hunk 6]
```diff
@@ -55,6 +57,10 @@ export class PersistentServerRepository implements ServerRepository {
     if (alreadyAddedServer) {
       throw new ServerAlreadyAdded(alreadyAddedServer);
     }
+    if (serverConfig.method &&
+        !PersistentServerRepository.SUPPORTED_CIPHERS.includes(serverConfig.method)) {
```
Comment (fortuna) [inline src/www/app/persistent_server.ts:61]: Provide a isSupportedCipher method instead.

File: src/www/app/main.ts [hunk 7]
```diff
@@ -52,9 +52,24 @@ function createServerRepo(
...
+        host: '192.0.2.1',
+        port: 123,
+        method: 'chacha20-ietf-poly1305'
+      });
+      repo.add({
+        name: 'Fake Unreachable Server',
+        host: '10.0.0.24',
+        port: 123,
+        method: 'chacha20-ietf-poly1305'
+      });
```
Comment (fortuna) [inline src/www/app/main.ts:72]: Maybe add one with invalid cipher?

File: cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java [hunk 8]
```diff
@@ -91,10 +91,11 @@ public boolean is(final String action) {
...
     SHADOWSOCKS_START_FAILURE(8),
-    CONFIGURE_SYSTEM_PROXY_FAILURE (9),
-    NO_ADMIN_PERMISSIONS (10),
-    UNSUPPORTED_ROUTING_TABLE (11),
-    SYSTEM_MISCONFIGURED (12);
+    CONFIGURE_SYSTEM_PROXY_FAILURE(9),
+    NO_ADMIN_PERMISSIONS(10),
+    UNSUPPORTED_ROUTING_TABLE(11),
+    SYSTEM_MISCONFIGURED(12),
+    UNSUPPORTED_CIPHER(13);
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java:98]: Done

File: src/www/messages/en.json (hunk 5 above)
Comment (mpmcroy) [reply src/www/messages/en.json:22]: Done

File: src/www/ui_components/server-card.html (hunk 4 above)
Comment (mpmcroy) [reply src/www/ui_components/server-card.html:180]: That makes sense.

The code duplication is bad. I chose this way because I don't know of how to share the logic between Typescript and Polymer 2 easily. Where could define it so that I can use it in both places?

File: src/www/app/main.ts (hunk 7 above)
Comment (mpmcroy) [reply src/www/app/main.ts:72]: Adding it would trigger our new error, so it wouldn't actually add it to the list.

File: src/www/ui_components/server-card.html (hunk 4 above)
Comment (fortuna) [reply src/www/ui_components/server-card.html:180]: As discussed on chat, the cipher validation shouldn't be in the UI. It should be in the Typescript application code. Let's move it there. The UI should just know how to display an error and disable connection.

Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: Thanks for the suggestion! Done.

File: src/www/app/main.ts (hunk 7 above)
Comment (mpmcroy) [reply src/www/app/main.ts:72]: To provide some more context here: 
We use a different approach to load servers from local storage. We parse the persisted JSON string instead of calling the add method.

File: resources/master_messages.json [hunk 9]
```diff
@@ -357,6 +357,16 @@
       }
     }
   },
+  "error_shadowsocks_unsupported_cipher": {
+    "message": "Shadowsocks Unsupported cipher",
+    "description": "The server you added uses cipher $cipher$, but Outline only supports AEAD ciphers. Please contact your server admin.",
```
Comment (fortuna) [inline resources/master_messages.json:362]: Maybe something like "The server you added uses cipher $cipher$, which is not safe. Please contact your server admin"

I think it's important to say it's not safe.

File: src/www/app/outline_server.ts [hunk 10]
```diff
@@ -21,6 +21,9 @@ import {Server} from '../model/server';
 import {PersistentServer} from './persistent_server';
 
 export class OutlineServer implements PersistentServer {
+  private static readonly SUPPORTED_CIPHERS =
```
Comment (fortuna) [inline src/www/app/outline_server.ts:24]: Add a comment with a reference to https://shadowsocks.org/en/spec/AEAD-Ciphers.html

File: src/www/app/persistent_server.ts [hunk 11]
```diff
@@ -56,6 +59,9 @@ export class PersistentServerRepository implements ServerRepository {
       throw new ServerAlreadyAdded(alreadyAddedServer);
     }
     const server = this.createServer(uuidv4(), serverConfig, this.eventQueue);
```
Comment (fortuna) [inline src/www/app/persistent_server.ts:61]: Let's check the `serverConfig.method` instead, before actually creating the server. It's not clear what the side-effects of createServer is. It's better to not have assumptions on that.

File: src/www/ui_components/server-card.html [hunk 12]
```diff
@@ -118,6 +118,13 @@
         color: var(--disabled-text-color);
         background: transparent;
       }
+
+      #unsupportedCipherMessage {
+        float: left;
```
Comment (fortuna) [inline src/www/ui_components/server-card.html:123]: This is not RTL-friendly. Do we really need to use `float`?
Maybe change to use [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) on card-actions instead. (and maybe get rid of the `text-align: right` there too)

File: src/www/ui_components/server-list.html [hunk 13]
```diff
@@ -49,6 +49,8 @@
         server-name="[[item.name]]"
         server-host="[[item.host]]"
         server-port="[[item.config.port]]"
+        server-cipher="[[item.config.method]]"
```
Comment (fortuna) [inline src/www/ui_components/server-list.html:52]: I think `server-cipher` doesn't exist

File: src/www/ui_components/server-card.html [hunk 14]
```diff
@@ -191,7 +200,7 @@
         },
         connectButtonDisabled: {
           type: Boolean,
-          computed: "_computeConnectButtonDisabled(state)",
+          computed: "_computeConnectButtonDisabled(state, usesUnsupportedCipher)",
```
Comment (fortuna) [inline src/www/ui_components/server-card.html:203]: I think this is not needed because the state will be DISCONNECTED anyway.

File: src/www/ui_components/server-card.html [hunk 15]
```diff
@@ -177,6 +185,7 @@
         serverId: String,
         serverHost: String,
         serverPort: Number,
+        usesUnsupportedCipher: Boolean,
```
Comment (fortuna) [inline src/www/ui_components/server-card.html:188]: Let's have a generic `errorMsg` property instead. This way we don't attach too much behavior to the value.
Let's also add a `enabled` property that is true by default and we can override to disable it. Or maybe just add a `DISABLED` state

This should simplify the UI.

File: src/www/ui_components/server-card.html [hunk 16]
```diff
@@ -225,8 +234,8 @@
           }[state] || this.localize("connect-button-label")
         );
       },
-      _computeConnectButtonDisabled: function(state) {
-        return state === "CONNECTING" || state === "DISCONNECTING";
+      _computeConnectButtonDisabled: function(state, usesUnsupportedCipher) {
+        return !usesUnsupportedCipher || state === "CONNECTING" || state === "DISCONNECTING";
```
Comment (fortuna) [inline src/www/ui_components/server-card.html:238]: We can revert this change, and add a check for a "disabled" state.

File: src/www/ui_components/server-list.html [hunk 17]
```diff
@@ -49,6 +49,8 @@
...
+        uses-unsupported-cipher="[[item.usesUnsupportedCipher]]"
```
Comment (fortuna) [inline src/www/ui_components/server-list.html:53]: This will changes with the suggestions above

File: src/www/app/outline_server.ts [hunk 18]
```diff
@@ -57,6 +60,11 @@ export class OutlineServer implements PersistentServer {
     return this.config.host;
   }
 
+  get usesUnsupportedCipher() {
```
Comment (fortuna) [inline src/www/app/outline_server.ts:63]: Let's add this to the Server model. It can be a a method. I think it makes it clearer it's calculated.

File: src/www/app/outline_server.ts [hunk 19]
```diff
@@ -57,6 +60,11 @@ export class OutlineServer implements PersistentServer {
...
+    return this.config.method === undefined ||
```
Comment (fortuna) [inline src/www/app/outline_server.ts:64]: Let's put this logic in a standalone function that can be reused.

File: package.json [hunk 20]
```diff
@@ -98,4 +98,4 @@
     "trailingComma": "es5",
     "printWidth": 120
   }
-}
+}
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply package.json]: Thanks, my IDE is being aggressive.

File: resources/master_messages.json (hunk 9 above)
Comment (mpmcroy) [reply resources/master_messages.json]: That sounds good. I tweaked the verbiage a bit.

File: src/www/app/persistent_server.ts (hunk 11 above)
Comment (mpmcroy) [reply src/www/app/persistent_server.ts:61]: That's a good point.

_createServer_ is platform specific, so we will need one implementation for each target environment. This was a shortcut to dedup the logic, but I see your concern.

File: src/www/ui_components/server-list.html (hunk 13 above)
Comment (mpmcroy) [reply src/www/ui_components/server-list.html]: Done.

File: src/www/ui_components/server-card.html (hunk 12 above)
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: Done.

File: src/www/app/persistent_server.ts (hunk 6 above)
Comment (mpmcroy) [reply src/www/app/persistent_server.ts]: Done

File: src/www/app/persistent_server.ts (hunk 11 above)
Comment (mpmcroy) [reply src/www/app/persistent_server.ts]: Oh, I see that I can just do this in OutlineServer instead.

File: src/www/ui_components/server-card.html (hunk 14 above)
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: I added the disabled property to address this

File: src/www/ui_components/server-card.html (hunk 15 above)
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: Done. I added both.

For the disabled property, I didn't flush it through all the way to the model. Instead I'm just checking for the existence of an error message :/ Lmk if we should

File: src/www/ui_components/server-card.html (hunk 16 above)
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: Done

File: src/www/ui_components/server-list.html (hunk 17 above)
Comment (mpmcroy) [reply src/www/ui_components/server-list.html]: Ack

File: src/www/app/outline_server.ts (hunk 18 above)
Comment (mpmcroy) [reply src/www/app/outline_server.ts]: Ack

File: src/www/app/outline_server.ts (hunk 19 above)
Comment (mpmcroy) [reply src/www/app/outline_server.ts]: Done

File: src/www/ui_components/server-card.html [hunk 21]
```diff
@@ -158,6 +165,7 @@
         <div class$="status-message [[state]]">[[statusMessage]]</div>
       </div>
       <div class="card-actions">
+        <div id="errorMessage">[[localize(errorMessage)]]</div>
```
[the remark this answers is not in the document]
Comment (mpmcroy) [reply src/www/ui_components/server-card.html]: Done

**Comment by LinxiFan** [comment]
This is weird - my shadowsocks addresses used to work on an older version of the iOS app. I updated to the latest version today, and now all of them show "Unsupported Cipher". I'm sure I could connect to those addresses before the upgrade. Help needed, thanks!
